Updates TPM buckets with a missing result key to default to 'n/a'

jasonrhodes opened the pull request:
Fixes #28838

## Summary

When `transaction.result` was missing, those transactions were grouped together with a key of `""`. We now default that bucket to have a key of `"N/A"` instead, so the graphs don't look broken.

<img width="740" alt="screen shot 2019-01-25 at 8 32 40 am" src="https://user-images.githubusercontent.com/159370/51748977-14e38780-207c-11e9-917a-33a135ad6b69.png">

jasonrhodes committed:
Updates tpm buckets with a missing result key to default to 'n/a'

sqren commented on x-pack/plugins/apm/server/lib/transactions/charts/get_timeseries_data/transform.ts:
```diff
@@ -64,6 +64,11 @@ export function getTpmBuckets(
+    if (key === '') {
+      key = 'N/A';
+    }
+
```
1) Have you given any thought to how `null` values should be treated? Normally buckets with `key=null` are stripped out but we include them under the key "transaction_result_missing":
https://github.com/elastic/kibana/blob/50ec75f800b91172c877b5aa0d6e446bb09780fb/x-pack/plugins/apm/server/lib/transactions/charts/get_timeseries_data/fetcher.ts#L142

Perhaps this logic should be kept together in the backend?

2) Localization ;)

3) Can we avoid the re-assignment and keep `key` immutable?
There is actually a tslint rule for this, that I think would make sense to enable `"no-parameter-reassignment": true`.

jasonrhodes replied:
>Perhaps this logic should be kept together in the backend?
Thoughts on how to do that besides changing the "missing" value to match this new value?

> Localization
Good catch, I see we have a localized `NOT_AVAILABLE_MESSAGE` in `public/constants/index.ts` that we can move to somewhere in `common` so it can be shared by server code.

>Can we avoid the re-assignment and keep key immutable?
I'm not as zealous about immutability as you are, I think, but if it makes you happy :)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9009/)

jasonrhodes committed:
Localizes "N/A", updates how missing values are handled

jasonrhodes commented:
Changes just pushed:
* Moved existing `NOT_AVAILABLE_LABEL` to the `common` directory and used it instead of hard-coding "N/A", updated all other references to use the correct path
* Removed mutation of `key` arg
* Changed the "missing" value for the result term aggregation from "transaction_result_missing" to just "", which will group all of the ES "missing" values with the empty string values and then all of them will be converted to "N/A" in the transform step, in one single "N/A" bucket (instead of 2 if we'd changed the "missing" value to "N/A" directly)
  * Because of this change, we no longer need to handle "transaction_result_missing" in the UI, so I removed that also

Last note: there seem to be some weird issues with ESLint not liking the "x-pack" import shortcut, but only in a few files. For now I've used `// eslint-disable-line import/no-extraneous-dependencies` to solve, but I've reached out to folks in @elastic/kibana-operations to see if there's a reason this changed.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9045/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9048/)

jasonrhodes commented:
> there seem to be some weird issues with ESLint not liking the "x-pack" import shortcut

Update: this shortcut pattern only works in `.ts`/`.tsx` files because tslint is ok with it, but eslint will complain. I'm not 100% sure it will actually work correctly in the build so I'll revert to "../.." paths in these files for now.

jasonrhodes committed:
Converts transaction and error group overview lists to TS, uses managed table component

jasonrhodes committed:
Fixed tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9146/)

jasonrhodes committed:
Fixes EUI-related snapshot update

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9205/)

sqren commented on x-pack/plugins/apm/public/components/app/ErrorGroupOverview/List/index.tsx:
```diff
@@ -0,0 +1,149 @@
+  ];
+
+  return (
+    <ManagedTable
```
Afaict this changes the behaviour, so errors are now sorted in-memory on the client - not in ES. 
If the user has 1000s of errors, and they try to sort by "occurrences" desc or "Latest occurrence" desc/asc they will now get incorrect results.

If we do decide to change this behaviour, we should also clean up all the server side sorting:
https://github.com/jasonrhodes/kibana/blob/e6590e7fcadc999f1786ddb4c33b62c3e4da62dd/x-pack/plugins/apm/server/lib/errors/get_error_groups.ts#L97-L107

It also doesn't look like `sortField` or `sortDirection` is used anymore:
https://github.com/jasonrhodes/kibana/blob/e6590e7fcadc999f1786ddb4c33b62c3e4da62dd/x-pack/plugins/apm/public/components/app/ErrorGroupOverview/List/index.tsx#L50

jasonrhodes replied:
Hmm, that's unfortunate that we can't consolidate this UX and then fix the client-side sorting issues once for everything, but we probably don't want to spread those issues before we have them fixed. I was hoping to avoid having to do lots of extra typing here. I'll revert.

jasonrhodes commented on x-pack/plugins/apm/server/lib/transactions/charts/get_timeseries_data/fetcher.ts:
```diff
@@ -105,52 +105,32 @@ export function timeseriesFetcher({
-            field: TRANSACTION_RESULT,
-            missing: 'transaction_result_missing'
-          },
+          terms: { field: TRANSACTION_RESULT, missing: '' },
```
Right, was trying to avoid having two “N/A” buckets.

jasonrhodes committed:
Reverts TS conversion and client-side sort change for ErrorGroupOverview

jasonrhodes commented:
@sqren changes to the ErrorGroupOverview List have been reverted.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9261/)

sqren commented on x-pack/plugins/apm/public/components/app/ErrorGroupOverview/index.tsx:
```diff
@@ -57,7 +54,7 @@ const ErrorGroupOverview: React.SFC<ErrorGroupOverviewProps> = ({
         urlParams={urlParams}
         render={({ data }) => (
-          <List urlParams={urlParams} items={data} location={location} />
+          <ErrorGroupList urlParams={urlParams} items={data} />
```
Should `location` not be passed to `ErrorGroupList`?

jasonrhodes replied:
Yup, and it needs to be re-passed here from the service tabs. Both fixed.

jasonrhodes committed:
Restores location passing for error group list

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9305/)
